securedrop_export/disk/cli.py

```diff
@@ -0,0 +1,423 @@
+
+    def get_luks_volume(self, device: str) -> Volume:
+        """
+        Given a string corresponding to a LUKS-partitioned volume, return a corresponding Volume
+        object.
+
+        If LUKS volume is already mounted, existing mountpoint will be preserved.
+        If LUKS volume is unlocked but not mounted, volume will be mounted at _DEFAULT_MOUNTPOINT.
+
+        If device is still locked, mountpoint will not be set. Once the decrpytion passphrase is
+        available, call unlock_luks_volume(), passing the Volume object and passphrase, to
+        unlock the volume.
+
+        Raise ExportException if errors are encountered.
+        """
+        try:
+            mapped_name = self._get_luks_name_from_headers(device)
+            logger.debug(f"Mapped name is {mapped_name}")
+
+            # Setting the mapped_name does not mean the device has already been unlocked.
+            luks_volume = Volume(
+                device_name=device,
+                mapped_name=mapped_name,
+                encryption=EncryptionScheme.LUKS,
+            )
+
+            # If the device has been unlocked, we can see if it's mounted and
+            # use the existing mountpoint, or mount it ourselves.
+            if os.path.exists(os.path.join("/dev/mapper/", mapped_name)):
+                return self.mount_volume(luks_volume)
+
+            # It's still locked
+            else:
+                return luks_volume
+
+        except ExportException:
+            logger.error("Failed to return luks volume")
+            raise
+
+    def unlock_luks_volume(self, volume: Volume, decryption_key: str) -> Volume:
+        """
+        Unlock a LUKS-encrypted volume.
+
+        Raise ExportException if errors are encountered during device unlocking.
+        """
+        if volume.encryption is not EncryptionScheme.LUKS:
+            logger.error("Must call unlock_luks_volume() on LUKS-encrypted device")
+            raise ExportException(sdstatus=Status.DEVICE_ERROR)
+
+        try:
+            logger.debug("Unlocking luks volume {}".format(volume.device_name))
+            p = subprocess.Popen(
+                [
+                    "sudo",
+                    "cryptsetup",
+                    "luksOpen",
+                    volume.device_name,
+                    volume.mapped_name,
+                ],
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+            logger.debug("Passing key")
+            p.communicate(input=str.encode(decryption_key, "utf-8"))
+            rc = p.returncode
+
+            if rc == 0:
+                return Volume(
+                    device_name=volume.device_name,
+                    mapped_name=volume.mapped_name,
+                    encryption=EncryptionScheme.LUKS,
+                )
+            else:
+                logger.error("Bad volume passphrase")
```

Comment:
From my recollection of an offline conversation: I think this was intentional, because it would be rendered as an error by the GUI. I don't remember if we were talking about this case specifically, but I remember thinking that I'd personally make log level decision based on the API of _the module_, and not what I know about how consumers decide to interpret its responses. (Because if the GUI decides to treat any given response as an unexpected error, then there is an opportunity right there to log it as such.)

In any case, I don't think there is a need to answer that question, or to establish a convention in the scope of this PR.